src/main/java/de/tum/in/www1/artemis/service/ComplaintResponseService.java

```diff
@@ -159,6 +159,43 @@ public ComplaintResponse createComplaintResponseRepresentingLock(Complaint compl
         return persistedComplaintResponse;
     }
 
+    /**
+     * Resolves a complaint by filling in the empty complaint response attached to it
+     *
+     * The empty complaint response acts as a lock. Only the creator of the empty complaint response and instructors can resolve empty complaint response as long as the lock
+     * is running. For lock duration calculation see: {@link ComplaintResponse#isCurrentlyLocked()}. These methods fill in the initial complaint response and either accepts
+     * or denies the associated complaint, thus resolving the complaint
+     *
+     * @param updatedComplaintResponse complaint response containing the information necessary for resolving the complaint
+     * @param complaintResponseId      complaint response id
+     * @return complaintResponse of resolved complaint
+     */
+    public ComplaintResponse resolveComplaint(ComplaintResponseUpdateDTO updatedComplaintResponse, Long complaintResponseId) {
+        validateComplaintResponseId(complaintResponseId);
+        ComplaintResponse complaintResponseFromDatabase = complaintResponseRepository.findByIdElseThrow(complaintResponseId);
+        validateComplaintResponseEmpty(complaintResponseFromDatabase);
+        Complaint originalComplaint = complaintRepository.findByIdElseThrow(complaintResponseFromDatabase.getComplaint().getId());
+        validateOriginalComplaintNotAnswered(originalComplaint);
+
+        User user = this.userRepository.getUserWithGroupsAndAuthorities();
+        validateUserPermissionAndLockStatus(originalComplaint, complaintResponseFromDatabase, user);
```

Comment:
Sorry for the mistake in my previous suggestion. Let's try this again. **This** is what I meant by ...
```suggestion
        // TODO: make this retrieval redundant by proper fetching
        ComplaintResponse complaintResponseFromDatabase = complaintResponseRepository.findByIdElseThrow(updatedComplaintResponse.getId());
        // TODO: make this retrieval redundant by proper fetching
        Complaint originalComplaint = complaintRepository.findByIdElseThrow(complaintResponseFromDatabase.getComplaint().getId());
        User user = this.userRepository.getUserWithGroupsAndAuthorities();

        validateUserPermissionAndLockStatus(originalComplaint, complaintResponseFromDatabase, user);
        validateComplaintResponseEmpty(complaintResponseFromDatabase);
        validateOriginalComplaintNotAnswered(originalComplaint);
```
This way, we first ensure that the authenticated user has the authorisation to access this resource for the purpose they want it to. Only then do we check the server entities for validity, which reveals information about the resource's state.

I also added the TODOs from below to this one. Consider extracting both to a helper method btw. You can also look for other calls that you can move in the same method.